presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java

```diff
@@ -0,0 +1,183 @@
+                if (compareResult < 0) {
+                    result = current;
+                }
+            }
+        }
+        return result;
+    }
+
+    public ListenableFuture<?> isBlocked()
+    {
+        if (mergeSources.stream().allMatch(source -> source.isBlocked().isDone())) {
+            return NOT_BLOCKED;
+        }
+        if (blocked != null && !blocked.isDone()) {
+            return blocked;
+        }
+        blocked = Futures.allAsList(mergeSources.stream().map(MergeSource::isBlocked).collect(Collectors.toList()));
+        return blocked;
+    }
+
+    public boolean isFinished()
+    {
+        return mergeSources.stream().allMatch(MergeSource::isFinished);
+    }
+
+    private static class MergeSource
+    {
+        private final PageSupplier pageSupplier;
+
+        private PageWithPosition currentPage;
+        private ListenableFuture<?> blocked;
+
+        public MergeSource(PageSupplier pageSupplier)
+        {
+            this.pageSupplier = requireNonNull(pageSupplier, "location is null");
+        }
+
+        public ListenableFuture<?> isBlocked()
+        {
+            if (currentPage != null && !currentPage.isFinished()) {
+                return NOT_BLOCKED;
+            }
+            if (blocked != null && !blocked.isDone()) {
+                return blocked;
+            }
+            blocked = pageSupplier.isBlocked();
+            return blocked;
+        }
+
+        public boolean isFinished()
+        {
+            return pageSupplier.isFinished() && (currentPage == null || currentPage.isFinished());
+        }
+
+        @Nullable
+        public PageWithPosition getPage()
+        {
+            if (currentPage != null && !currentPage.isFinished()) {
+                return currentPage;
+            }
+
+            checkState(blocked != null, "blocked is null");
+            if (!isBlocked().isDone()) {
+                return null;
+            }
+
+            if (pageSupplier.isFinished()) {
+                return null;
+            }
+
+            Page page = pageSupplier.poolPage();
+            currentPage = new PageWithPosition(page);
+            return currentPage;
+        }
+    }
+
+    public static class PageWithPosition
+    {
+        private final Page page;
+        private int position = 0;
+
+        public PageWithPosition(Page page)
+        {
+            this.page = requireNonNull(page, "page is null");
+        }
+
+        public Page getPage()
+        {
+            return page;
+        }
+
+        public int getPosition()
+        {
+            checkPosition();
+            return position;
+        }
+
+        public void incrementPosition()
+        {
+            checkPosition();
+            position++;
+        }
+
+        public boolean isFinished()
+        {
+            return position == page.getPositionCount();
```

Comment:
This make some sense, but i believe it is too paranoid. We do checkstate in increment.